mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java

```diff
@@ -0,0 +1,301 @@
+import com.google.common.base.Predicate;
+import com.google.common.base.Throwables;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableList;
+
+public class StoreMessageIdManager implements MessageIdManager {
+
+    private static final Function<MailboxMessage, MetadataWithMailbox> EXTRACT_METADATA_FUNCTION = new Function<MailboxMessage, MetadataWithMailbox>() {
+        @Override
+        public MetadataWithMailbox apply(MailboxMessage mailboxMessage) {
+            return new MetadataWithMailbox(new SimpleMessageMetaData(mailboxMessage), mailboxMessage.getMailboxId());
+        }
+    };
+    public static final Function<MessageAttachment, Attachment> TRANSFORM_TO_ATTACHEMENT = new Function<MessageAttachment, Attachment>() {
+        @Override
+        public Attachment apply(MessageAttachment input) {
+            return input.getAttachment();
+        }
+    };
+
+    private final MailboxSessionMapperFactory mailboxSessionMapperFactory;
+    private final MailboxEventDispatcher dispatcher;
+    private final MessageId.Factory messageIdFactory;
+    private final MessageParser messageParser;
+    private final PermanentFlagsProvider permanentFlagsProvider;
+    private final QuotaManager quotaManager;
+    private final QuotaRootResolver quotaRootResolver;
+
+    public StoreMessageIdManager(MailboxSessionMapperFactory mailboxSessionMapperFactory, MailboxEventDispatcher dispatcher,
+                                 MessageId.Factory messageIdFactory, MessageParser messageParser, PermanentFlagsProvider permanentFlagsProvider,
+                                 QuotaManager quotaManager, QuotaRootResolver quotaRootResolver) {
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+        this.dispatcher = dispatcher;
+        this.messageIdFactory = messageIdFactory;
+        this.messageParser = messageParser;
+        this.permanentFlagsProvider = permanentFlagsProvider;
+        this.quotaManager = quotaManager;
+        this.quotaRootResolver = quotaRootResolver;
+    }
+
+    @Override
+    public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, MessageId messageId, MailboxSession mailboxSession) throws MailboxException {
+        MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
+        Map<MailboxId, UpdatedFlags> updatedFlags = messageIdMapper.setFlags(newState, replace, messageId);
+        for (Map.Entry<MailboxId, UpdatedFlags> entry : updatedFlags.entrySet()) {
+            dispatchFlagsChange(mailboxSession, entry);
+        }
+    }
+
+    private void dispatchFlagsChange(MailboxSession mailboxSession, Map.Entry<MailboxId, UpdatedFlags> entry) throws MailboxException {
+        Mailbox mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxById(entry.getKey());
+        if (!entry.getValue().getNewFlags().equals(entry.getValue().getOldFlags())) {
```

Comment:
this line only involve UpdatedFlags values, then it probably belongs to UpdatedFlags